query/src/sql/optimizer/heuristic/decorrelate.rs

```diff
@@ -12,224 +12,725 @@
+                let op = subquery.compare_op.as_ref().unwrap().clone();
+                // Make <child_expr op right_condition> as other_conditions even if op is equal operator.
+                // Because it's not null-safe.
+                let other_conditions = vec![Scalar::ComparisonExpr(ComparisonExpr {
+                    op,
+                    left: Box::new(child_expr),
+                    right: Box::new(right_condition),
+                    return_type: Box::new(NullableType::new_impl(BooleanType::new_impl())),
+                })];
+                let mut metadata = self.metadata.write();
+                let marker_index = metadata.add_column(
+                    "marker".to_string(),
+                    NullableType::new_impl(BooleanType::new_impl()),
+                    None,
+                );
+                let mark_join = LogicalInnerJoin {
+                    left_conditions,
+                    right_conditions,
+                    other_conditions,
+                    join_type: JoinType::Mark,
+                    marker_index: Some(marker_index),
+                    from_correlated_subquery: true,
+                }
+                .into();
+                Ok((
+                    SExpr::create_binary(mark_join, flatten_plan.clone(), left.clone()),
+                    UnnestResult::MarkJoin { marker_index },
+                ))
+            }
+            _ => unreachable!(),
+        }
     }
-    // Add project for join keys
-    let mut used_columns = join
-        .right_conditions
-        .iter()
-        .fold(ColumnSet::new(), |v, acc| {
-            v.union(&acc.used_columns()).cloned().collect()
-        });
-    used_columns = used_columns
-        .union(
-            &join
-                .other_conditions
-                .iter()
-                .fold(ColumnSet::new(), |v, acc| {
-                    v.union(&acc.used_columns()).cloned().collect()
-                }),
-        )
-        .cloned()
-        .collect();
-    used_columns = used_columns
-        .difference(
-            &used_columns
-                .intersection(&input_prop.output_columns)
-                .cloned()
-                .collect(),
-        )
-        .cloned()
-        .collect();
-
-    right_child = SExpr::create_unary(
-        Project {
-            columns: used_columns,
+
+    fn flatten(&mut self, plan: &SExpr, correlated_columns: &ColumnSet) -> Result<SExpr> {
+        let rel_expr = RelExpr::with_s_expr(plan);
+        let prop = rel_expr.derive_relational_prop()?;
+        if prop.outer_columns.is_empty() {
+            // Construct a LogicalGet plan by correlated columns.
+            // Finally generate a cross join, so we finish flattening the subquery.
+            let mut metadata = self.metadata.write();
+            let table_index = metadata
+                .table_index_by_column_indexes(correlated_columns)
+                .unwrap();
```

Comment:
Does this work if we have such a query:
```sql
select * from (select a + 1 as a from t) as t where exists(select * from t as t1 where t1.a < t.a)
```

Reply:
I think maybe you can copy the left child of dependent join to replace the `LogicalGet` here, since we don't have mechanism to reuse the output of a processor yet.

Reply:
Yes
```sql
mysql> create table t(a int);
Query OK, 0 rows affected (0.03 sec)

mysql> insert into t values(1),(2);
Query OK, 0 rows affected (0.03 sec)

mysql> select * from (select a + 1 as a from t) as t where exists(select * from t as t1 where t1.a < t.a);
+------+
| a    |
+------+
|    2 |
|    3 |
+------+
2 rows in set (0.04 sec)
Read 4 rows, 16.00 B in 0.004 sec., 922.28 rows/sec., 3.60 KiB/sec.
```

Reply:
> I think maybe you can copy the left child of dependent join to replace the `LogicalGet` here, since we don't have mechanism to reuse the output of a processor yet.

In fact, I want to add a new `rel_op` later which contains columns from left child of dependent join and will eliminate duplicate rows.

Reply:
> In fact, I want to add a new `rel_op` later which will get distinct columns from left child of dependent join.

We can do that later. In DuckDB, they have `DelimScan` and `DelimJoin` to do the work, which is heavily dependent on the runtime.

Reply:
> We can do that later. In DuckDB, they have `DelimScan` and `DelimJoin` to do the work, which is heavily dependent on the runtime.

Ok

Reply:
> Well, seems this can be converted to semi join.
> 
> What about:
> 
> ```sql
>  create table t(a int);
> insert into t values(1), (2);
> select * from (select a + 1 as a from t) as t where (select sum(a) from t as t1 where t1.a < t.a) = 1;
> ```

I get you, will give a try to copy the left child of dependent join.

Reply:
Derived an issue: https://github.com/datafuselabs/databend/issues/6688, will support the case in the next PR.